arrow-libs/fx/arrow-fx-coroutines/src/commonMain/kotlin/arrow/fx/coroutines/Racing.kt

```diff
@@ -0,0 +1,129 @@
+import kotlin.coroutines.EmptyCoroutineContext
+import kotlin.coroutines.cancellation.CancellationException
+
+/**
+ * A DSL that allows racing many `suspend` functions in parallel against each-other,
+ * it yields a final result of [A] based on the first function that yields a result.
+ * A racer can yield a result based on [RacingScope.race], or [RacingScope.raceOrFail].
+ *
+ * [RacingScope.race] will call [handleException] in case of an exception,
+ * and then await **another successful result** but not cancel the race. Whilst [RacingScope.raceOrFail] will cancel the race,
+ * and rethrow the exception that occurred and thus cancel the race and all participating racers.
+ *
+ * ```kotlin
+ * suspend fun winner(): String = racing {
+ *   race { delay(1000); "Winner" }
+ *   race { throw RuntimeException("Loser") }
+ * } // Winner (logged RuntimeException)
+ *
+ * suspend fun winner(): String = racing {
+ *   race { delay(1000); "loser" }
+ *   raceOrFail { throw RuntimeException("Loser") }
+ * } // RuntimeException
+ * ```
+ *
+ * **Important:** a racing program with no racers will hang forever.
+ * ```kotlin
+ * suspend fun never(): Nothing = racing { }
+ * ```
+ *
+ * @param handleException handle any exception that occurred in [RacingScope.race],
+ * by default it [Throwable.printStackTrace].
+ * @param block the body of the DSL that describes the racing logic
+ * @return the winning value of [A].
+ */
+public suspend fun <A> racing(
+  handleException: ((context: CoroutineContext, exception: Throwable) -> Unit)? = null,
+  block: RacingScope<A>.() -> Unit,
+): A = coroutineScope {
+  val exceptionHandler = handleException ?: defaultExceptionHandler()::handleException
+  select {
+    val scope = SelectRacingScope(this@select, this@coroutineScope, exceptionHandler)
+    block(scope)
+    // TODO add this check??
+    require(scope.racers.value.isNotEmpty()) { "A racing program with no racers can never yield a result." }
+  }
+}
+
+public interface RacingScope<A> {
+  public suspend fun race(
+    context: CoroutineContext = EmptyCoroutineContext,
+    block: suspend CoroutineScope.() -> A
+  )
+
+  public suspend fun raceOrFail(
+    context: CoroutineContext = EmptyCoroutineContext,
+    block: suspend CoroutineScope.() -> A
+  )
+}
+
+private class SelectRacingScope<A>(
+  private val select: SelectBuilder<A>,
+  private val scope: CoroutineScope,
+  private val handleException: (context: CoroutineContext, exception: Throwable) -> Unit
+) : RacingScope<A>, CoroutineScope by scope {
+  val racers: Atomic<List<Deferred<A>>> = Atomic(emptyList())
+
+  override suspend fun raceOrFail(
+    context: CoroutineContext,
+    block: suspend CoroutineScope.() -> A
+  ) {
+    /* First we create a lazy racer,
+     * and we add it in front of the existing racers such that we maintain correct order.
+     * After we've successfully registered the racer, we check for race conditions,
+     * and 'start' racing.
+     */
+    val racer = scope.async(
```

Comment:
Nit: we're already implementing `CoroutineScope` by delegation, so I'd rather this be a call on this directly.
```suggestion
    val racer = async(
```

Reply:
Without the `scope` it results in "ambiguous coroutine context", so I am going to revert extending `CoroutineScope` but keep it as a `private val`.